client/text.cpp

```diff
@@ -1731,41 +1731,132 @@ QString text_happiness_cities(const struct city *pcity)
     }
-    str += QString(PL_("Each military unit makes %1 "
-                       "unhappy citizen content.",
-                       "Each military unit makes %1 "
-                       "unhappy citizens content.",
-                       mleach))
-               .arg(QString::number(mleach))
-           + qendl();
-  } else if (uhcfac > 0) {
-    str += QString(_("Military units in the field may cause unhappiness. "))
-           + qendl();
+
+    str += QStringLiteral(" ");
+    str += QString(PL_("Each of them makes %1 unhappy citizen content.",
+                       "Each of them makes %1 unhappy citizens content.",
+                       martial_law_each))
+               .arg(martial_law_each);
+
+    str += QStringLiteral("</p><p>");
+
+    int count = 0;
+    unit_list_iterate(pcity->tile->units, punit)
+    {
+      if ((count < martial_law_max || martial_law_max == 0)
+          && is_military_unit(punit)
+          && unit_owner(punit) == city_owner(pcity)) {
+        count++;
+      }
+    }
+    unit_list_iterate_end;
+
+    str +=
+        QString(PL_("%1 military unit in this city imposes the martial law.",
+                    "%1 military units in this city impose the martial law.",
+                    count))
+            .arg(count);
+    str += QStringLiteral(" ");
+    str += QString(PL_("<b>%1 citizen</b> is made happier as a result.",
+                       "<b>%1 citizens</b> are made happier as a result.",
+                       count * martial_law_each))
+               .arg(count * martial_law_each);
+
+    str += QStringLiteral("</p>");
+  }
+
+  /*
+   * Second part: military unhappiness
+   *
+   * Had to say anything here because this is unit dependent (and I don't
+   * want to build a list of unit types or so, this is shown elsewhere).
+   */
+  str += QStringLiteral("<p>");
+
+  auto pplayer = city_owner(pcity);
+  int unhappy_factor = get_player_bonus(pplayer, EFT_UNHAPPY_FACTOR);
+  if (unhappy_factor == 0) {
+    str += _("Military units have no happiness effect.");
+    return str + QStringLiteral("</p>");
+  }
+
+  str += _("Military units in the field may cause unhappiness.");
+
+  // Special rule: ignore the first N military unhappy
+  int made_content = get_city_bonus(pcity, EFT_MAKE_CONTENT_MIL);
+  if (made_content > 0) {
+    str += QStringLiteral(" ");
+    str += QString(_("The unhappiness caused by military units is reduced "
+                     "by %1 with respect to its normal value."))
+               .arg(made_content);
+  }
+
+  // Special rule: ignore N military unhappy per unit
+  // Let's hope ruleset authors don't combine them too often, the sentences
+  // are very similar.
+  int made_content_each = get_city_bonus(pcity, EFT_MAKE_CONTENT_MIL_PER);
+  if (made_content_each > 0) {
+    str += QStringLiteral(" ");
+    str += QString(_("The unhappiness caused by each military unit is "
+                     "reduced by %1 with respect to its normal value."))
```

Comment:
This also feels clunky. Perhaps something like "With respect to normal, %1 fewer citizens will be unhappy per each military unit."?

Reply:
Approved!